object destructuring

zdenko opened the pull request:
@GeoffreyBooth I've made some improvements regarding the object destructuring assignment and rest elements.
Basically, implement proposal for [rest properties for ECMAScript](https://github.com/tc39/proposal-object-rest-spread) in CS: `{a, b, c...} = x`

Since this proposal isn't yet at stage-4, CS doesn't compile to ES.
In my PR I catch rest element and assign remaining values to it:
`{a, b, c...} = x` compiles to:
`a = x.a, b = x.b, c = extractObjectWithoutKeys(x, ['a', 'b'])`

Multiple rest elements are disallowed. ES also requires the rest element to be the last, so compiler currently throws an error.  
IMHO, CS should allow rest element anywhere, just like for arrays.
Later, when proposal reaches stage-4 we could implement similar conversion as you did for the function parameters, and ensure rest element is the last in the compiled JS.

GeoffreyBooth committed:
Don’t confuse the syntax highlighter

GeoffreyBooth committed:
Don’t confuse the syntax highlighter

lydell committed:
Comment Assign::compilePatternMatch a bit

lydell committed:
Comment Assign::compilePatternMatch a bit

GeoffreyBooth committed:
Merge branch 'destructuring-comments' of github.com:lydell/coffee-script into destructuring

GeoffreyBooth committed:
Assignment expressions in conditionals are a bad practice

GeoffreyBooth committed:
Assignment expressions in conditionals are a bad practice

GeoffreyBooth committed:
Merge branch '2' into destructuring

GeoffreyBooth committed:
Rename `wrapInBraces` to `wrapInParentheses`, to set the stage for future `wrapInBraces` that uses `{` and `wrapInBrackets` that uses `[`

GeoffreyBooth committed:
Rename `wrapInBraces` to `wrapInParentheses`, to set the stage for future `wrapInBraces` that uses `{` and `wrapInBrackets` that uses `[`

GeoffreyBooth committed:
Correct comment

Commit:
object destructuring

Commit:
object destructuring

Commit:
Allow custom position of the rest element.

Commit:
Allow custom position of the rest element.

zdenko commented:
I removed the requirement for the position of the rest element.
CS will always compile it as the last element.

Currently, `{a, b, c...} = x` and `{a, c..., b} = x` are compiled to 
`a = x.a, b = x.b, c = extractObjectWithoutKeys(x, ['a', 'b'])`.

When ES proposal hits the stage-4 it will compile to `{a, b, ...c} = x`

GeoffreyBooth commented on .gitignore:
```diff
@@ -8,3 +8,4 @@ test/*.js
 parser.output
 /node_modules
 npm-debug.log*
+yarn.lock
```
Should this be in here?

GeoffreyBooth commented on src/nodes.coffee:
```diff
@@ -3028,6 +3063,20 @@ UTILITIES =
   "
+  
+  # copy object properties excluding the list of keys
+  extractObjectWithoutKeys: (o) -> "
```
CS2 assumes ES2015+. Do we need this function? Is there no ES function like `Object.keys` that can do some or all of this work for us?

I’m not opposed to the function if we need it; just trying to avoid helpers unless we need them.

GeoffreyBooth commented on test/assignment.coffee:
```diff
@@ -227,7 +227,64 @@ test "destructuring assignment with objects and splats", ->
+test "destructuring assignment with objects and splats: ES2015", ->
+  obj = {a:1, b:2, c:3, d:4, e:5}
+  
+  # throws (-> CoffeeScript.compile "{a, r..., b} = x"), null, "rest element must be the last"
```
Please remove commented-out code.

GeoffreyBooth commented on test/assignment.coffee:
```diff
@@ -227,7 +227,64 @@ test "destructuring assignment with objects and splats", ->
+  eq w, a
+  eq r['b'], obj.b
+  
+  # throws (-> CoffeeScript.compile "{a, b: {r..., c}} = x"), null, "rest element must be the last"
```
Ditto. Please also follow the style of the file, two empty lines between tests.

GeoffreyBooth commented on src/grammar.coffee:
```diff
@@ -189,6 +189,9 @@ grammar =
   # the ordinary **Assign** is that these allow numbers and strings as keys.
   AssignObj: [
     o 'ObjAssignable',                          -> new Value $1
+    
```
Please follow the style of the file, without empty lines here.

GeoffreyBooth reviewed (changes requested):
I haven’t reviewed the guts of what you’ve done, just some superficial comments on the stuff around it. Bravo though for implementing this feature without needing to output the stage-3 syntax. Perhaps you should add some comments like “When XXX is finalized, remove these lines” or somesuch so that we know what to revise once the syntax gets final approval?

zdenko commented:
I've made some progress yesterday and currently, I'm able to compile:

```javascript
{a, b:z, c} = {a:1, b:2, c:3, d:4, e:5}

obj1 = {f:1, g:2, h:3}
{f, g} = obj1
```

into

```javascript
var obj1;

({a, b:z, c} = {a:1, b:2, c:3, d:4, e:5});

obj1 = {f:1, g:2, h:3};
({f, g} = obj1);
```

and with rest element

```javascript
{a, b:z, c, r...} = {a:1, b:2, c:3, d:4, e:5}

obj1 = {f:1, g:2, h:3}
{f, q...} = obj1
```

into

```javascript
var obj1, q, r, ref;

ref = {a: 1, b: 2, c: 3, d: 4, e: 5}, ({a, b: z, c} = ref), 
r = Object.keys(ref).filter((k) => ['a','b','c'].indexOf(k) == -1).reduce((a, c) => (a[c] = ref[c], a), {});

obj1 = {f: 1, g: 2, h: 3};
({f} = obj1), 
q = Object.keys(obj1).filter((k) => ['f'].indexOf(k) == -1).reduce((a, c) => (a[c] = obj1[c], a), {});
```

I'm stuck with declarition of the variables from the object at the top of the scope.
Any hints or suggestions how to approach this?

So this:

```javascript
obj1 = {f:1, g:2, h:3}
{f, g} = obj1
```

should compile to

```javascript
var obj1, f, g;

obj1 = {f:1, g:2, h:3};
({f, g} = obj1);
```

GeoffreyBooth commented:
Take a look at https://github.com/jashkenas/coffeescript/pull/4478. I was working last night too and I got that part working. Merge my branch into yours. 

The declaration happens automatically when you do `new Assign`.

Commit:
rest element in object destructuring

Commit:
rest element in object destructuring

Commit:
fix string interpolation

Commit:
merge

Commit:
merging

Commit:
fixing splats in object literal

zdenko commented:
So, with the latest commit, I covered simple and nested destructuring assignment: 

```javascript
obj = {a:1, b:2, c:3, d:4, e:5}
{a, b:z, d, r...} = obj

a1={}; b1={}; c1={}; d1={}
obj = {
    a: a1
    b: {
      'c': {
        d: {
          b1
          e: c1
          f: d1
        }
      }
    }
    b2: {b1, c1}
}
{a:w, 'b':{c:{d:{b1:bb, r1...}}}, r2...} = obj
```

I added the `compileObjectDestruct` method to the `Assign` class, in which I basically traverse thru the object, remove splats and create a list of assignments which are returned back to the `compileNode`.
For that reason, I also added `answers` variable in the `compileNode` where possible splat assignments are stored.
When ES proposal hits stage-4, it should be pretty simple to remove `compileObjectDestruct` method and `answers` variable.

I don't check for the position of the splat. Currently, it is removed and the compiled code is correct, and afterward, it can be always compiled as the last element: `{a, r..., b}`  => `{a, b, ...r}`

I guess the next step could be splat in function parameters, which could compile similar as array splats:
```javascript
fn = ({a, b, r...}) -> {}
```

```javascript
fn = function(arg) {
   var a, b, r;
   ({a, b} = arg), r = Object.keys(arg)......
}
```

GeoffreyBooth commented:
Well . . . what _about_ function parameters, though? As we've been discussing in https://github.com/jashkenas/coffeescript/pull/4478, parameters don't get passed through `Assign`.

Commit:
Rest element in parameter destructuring

zdenko commented:
I've made attempt to enable splats in function parameters destructuring.

```javascript
fn = ({a, b, r...}) -> {}
```

```javascript
fn = function(arg) {
   var arg, a, b, r;
   ({a, b} = arg), r = Object.keys(arg)......
}
```

When parsing parameters I'm checking for possible splats. If they exists, I pass the object through `Assign`, push results to the function body and create required variables in function scope.

zdenko commented:
I saw you closed this PR. Any feedback on last commits?
Should I continue with the work? 
I'm almost done with the spread properties
https://github.com/tc39/proposal-object-rest-spread/blob/master/Spread.md

GeoffreyBooth commented:
Sorry I didn’t intentionally close it, it got closed because my `destructuring` branch got merged (and then deleted). We need to change the target of this PR to the jashkenas:coffeescript `2` branch.

GeoffreyBooth commented:
Looks like I can’t modify the merge target. Can you edit the PR to point to the `2` branch?

And yes please continue working, I’m eager to get this merged in! Thanks!

zdenko commented:
Did you perhaps check the function parameters destructuring (the last commit)?
Is the approach acceptable?
And sorry, I wasn't paying full attention to the discussion in [jashkenas#4478](https://github.com/jashkenas/coffeescript/pull/4478) 
Bussy days. Plus, I just dived into this without a deeper overview of CS internals.
